## ■ 다트게임 (9주차// Jan.13th)

## ● 문제1. input 과 split를 이용하시오!

In [ ]:
res = input("입력 : ")
print(res.split())  # 띄어쓰기를 기준으로 리스트에 요소가 들어간다.

## ● 문제2. 공백을 중간에 넣지 않고 리스트화 되게 하기

In [1]:
a = input('입력하시오 ~')

list_a = [i for i in a]
print(list_a)

입력하시오 ~1234
['1', '2', '3', '4']


## ● 문제3. 문제2번 코드 재확인.

In [3]:
a = input('입력하시오 ~')

list_a = [i for i in a]
print(list_a)

입력하시오 ~1S2D*3T
['1', 'S', '2', 'D', '*', '3', 'T']


## ● 문제4. 아래와 같이 실행 되게 하시오!(리스트 변수 3개 생성)

In [4]:
num = [str(i) for i in range(11)]
bonus = ['S', 'D', 'T']
option = ['*', '#']

print(num)
print(bonus)
print(option)

['0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '10']
['S', 'D', 'T']
['*', '#']


## ● 문제5. 1S2D3T에서 1,2,3 숫자만 출력해보기

In [5]:
num = ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '10']
bonus = ['S', 'D', 'T']
option = ['*', '#']


res = input("입력 : ")
a = [i for i in res if i in num]

print(a)

입력 : 1s2d3t
['1', '2', '3']


## ● 문제6. bonus 리스트에 S의 인덱스 번호를 출력하시오!

In [6]:
bonus = ['S', 'D', 'T']


print(bonus.index('S'))

0


## ● 문제7. bonus 리스트에서 S의 인덱스 번호를 1로 출력하게 하고 D는 2로 출력하게 하고 T는 3으로 출력되게 하시오!

In [8]:
bonus = ['S', 'D', 'T'] # S는 1제곱 , D는 2제곱 , T는 3제곱 해주기 위해서 이 코드를 짜고있음.

print(bonus.index('S')+1)

1


## ● 문제8.  2^3을 파이썬으로 출력하시오!

In [9]:
print(2**3)

8


## ● 문제9. 1S2D3T에서 숫자는 추출했는데 S와 D와 T를 bonus 변수를 이용해서 추출해서 아래와 같이 S는 1로 D는 2로 T는 3으로 출력되게 하시오!

In [10]:
input_data = input("입력하시오 : ")
dart = [i for i in input_data]

num = ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '10']
bonus = ['S', 'D', 'T']
option = ['*', '#']

dart_num = [i for i in dart if i in num]
dart_bonus = [bonus.index(i)+1 for i in dart if i in bonus]

print(dart_bonus)

입력하시오 : 1S2D3T
[1, 2, 3]


## ● 문제10. 위의 코드에서 10S2D3T을 입력했을 때 아래와 같이 출력 되게 하시오!

In [11]:
input_data = input("입력 : ").replace('10', 'k')
dart = [i for i in input_data] 

# 현재 한 글자씩 잘라서 리스트의 요소로 넣기 때문에 '10'으로 들어가지 않고 '1', '0' 으로 들어간다.
# 따라서, 10을 replace를 이용해서 'k' 라는 한글자로 표현함.
# 예제2번 때문에 필요함.

num = ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'k']
bonus = ['S', 'D', 'T']
option = ['*', '#']

dart_num = ['10' if i == 'k' else i  for i in dart if i in num] # 'dart' 리스트안에 'k'가 있으면 10으로 바꿔줌.
dart_bonus = [bonus.index(i)+1 for i in dart if i in bonus]

print(dart_num)

입력 : 10S2D3T
['10', '2', '3']


## ● 문제11. 위의 코드를 이용해서 10S4D5T 를 입력하면 아래의 식으로 변환되게 하시오!

10 ** 1 + 4 ** 2 + 5 ** 3

In [12]:
input_data = input("입력 : ").replace('10', 'k') #10S4D5T
dart = [i for i in input_data] 


num = ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'k']
bonus = ['S', 'D', 'T']
option = ['*', '#']


dart_num = [ 10 if i == 'k' else int(i)  for i in dart if i in num] # 'dart' 리스트안에 'k'가 있으면 10으로 바꿔줌.
dart_bonus = [bonus.index(i)+1 for i in dart if i in bonus]

print(dart_num)
print(dart_bonus)

for i in range(len(dart_num)):
    print(dart_num[i], "**", dart_bonus[i], "=", dart_num[i] ** dart_bonus[i])

입력 : 10S4D5T
[10, 4, 5]
[1, 2, 3]
10 ** 1 = 10
4 ** 2 = 16
5 ** 3 = 125


In [ ]:
# 결과를 한줄로 표현하기
# 10**1 + 4**2 + 5**3

input_data = input("입력 : ").replace('10', 'k') #10S2D3T
dart = [i for i in input_data] 


num = ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'k']
bonus = ['S', 'D', 'T']
option = ['*', '#']


dart_num = [ 10 if i == 'k' else int(i)  for i in dart if i in num] # 'dart' 리스트안에 'k'가 있으면 10으로 바꿔줌.
dart_bonus = [bonus.index(i)+1 for i in dart if i in bonus]

print(dart_num)
print(dart_bonus)

str1 = ''
for i in range(len(dart_num)):
    str1 += str(dart_num[i]**dart_bonus[i]) + '+' # 문자열 합치기

# 10+4+27+

print(str1.rstrip('+')) # rstrip으로 맨 오른쪽에있는 '+' 를 없애준다.
# 10+4+27

## 문제12. 위의 코드에 eval 을 이용해서 위의 출력된 결과인 10+16+125 가 계산되게 하시오!

In [ ]:
# 결과를 한줄로 표현하기
# 10**1 + 4**2 + 5**3

input_data = input("입력 : ").replace('10', 'k') #10S2D3T
dart = [i for i in input_data] 


num = ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'k']
bonus = ['S', 'D', 'T']
option = ['*', '#']


dart_num = [ 10 if i == 'k' else int(i)  for i in dart if i in num] # 'dart' 리스트안에 'k'가 있으면 10으로 바꿔줌.
dart_bonus = [bonus.index(i)+1 for i in dart if i in bonus]

print(dart_num)
print(dart_bonus)

str1 = ''
for i in range(len(dart_num)):
    str1 += str(dart_num[i]**dart_bonus[i]) + '+' # 문자열 합치기

# 10+4+27+

print(eval(str1.rstrip('+'))) # rstrip으로 맨 오른쪽에있는 '+' 를 없애준다.
# 10+4+27

### 주의사항 : 예제2번, 예제4번 확인해볼것!

### 힌트 : 스타상(*) 을 적용하기 위해서 전에 했던 값을 저장할 필요도있다.

### if, elif, append, index, replace, 문자열더하기 , 빈리스트 하나 만들기 등등

In [ ]:
# 카카오 문자열 처리 알고리즘 일반 version1

d = '1T2D3D#'
n = [str(i) for i in range(10)]
a = ['S', 'D', 'T']
s = ['*', '#']

li = []
answer = 0
tmp1 = ''

for i in range(len(d)):
    if d[i] in n:
        tmp1 += d[i]
    
    elif d[i] in a:
        num = int(tmp1)
        
        if d[i] == 'S':
            tmp2 = num**1
            
        elif d[i] == 'D':
            tmp2 = num**2

        elif d[i] == 'T':
            tmp2 = num**3       
        tmp1 = ''
        li.append(tmp2)
        
    elif d[i] == '*':
        if len(li) >= 2:
            li[len(li)-2] = li[len(li)-2] * 2
            li[len(li)-1] = li[len(li)-1] * 2
        else:
            li[len(li)-1] = li[len(li)-1] * 2
        
    elif d[i] == '#':
        li[len(li)-1] = li[len(li)-1] * (-1)
        
for i in li:
    answer += i
    
print(answer)

In [ ]:
def solution(dartResult):
    answer = 0
    n = [str(i) for i in range(10)]
    a = ['S', 'D', 'T']
    li = []

    tmp1 = ''
    for i in range(len(dartResult)):
        if dartResult[i] in n:
            tmp1 += dartResult[i]
            
        elif dartResult[i] in a:
            num = int(tmp1)
            if dartResult[i] == 'S':
                result = num**1
                
            elif dartResult[i] == 'D':
                result = num**2
    
            elif dartResult[i] == 'T':
                result = num**3   
                
            tmp1 = ''
            li.append(result)
            
        elif dartResult[i] == '*':
            if len(li) >= 2:
                li[len(li)-2] = li[len(li)-2] * 2
                li[len(li)-1] = li[len(li)-1] * 2
            else:
                li[len(li)-1] = li[len(li)-1] * 2
            
        elif dartResult[i] == '#':
            li[len(li)-1] = li[len(li)-1] * (-1)
            
    for i in li:
        answer += i
        
    return answer

dartResult = '1S2D*3T'
print(solution(dartResult))